[MPS] Add MacOS 14 runtime check

malfet opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #115514
* #115554
* #115513
* __->__ #115512

Prerequisite for adding more complex type support and FFT operation

Check using `conjugateWithTensor:name:` selector defined as follows
```objc
/// Returns the complex conjugate of the input tensor elements.
///
/// - Parameters:
///   - tensor: The input tensor.
///   - name: An optional string which serves as an identifier for the operation..
/// - Returns: A valid `MPSGraphTensor` object containing the elementwise result of the applied operation.
-(MPSGraphTensor *) conjugateWithTensor:(MPSGraphTensor *) tensor
                                   name:(NSString * _Nullable) name
MPS_AVAILABLE_STARTING(macos(14.0), ios(17.0), tvos(17.0))
MPS_SWIFT_NAME( conjugate(tensor:name:) );
```

- Rename `isOnMacOS13orNewer(unsigned minor)` hook to `isOnMacOSorNewer(major, minor)`
- Replace `torch._C.__mps_is_on_macos_13_or_newer` with `torch._C._mps_is_on_macos_or_newer`
- Add `torch.backends.mps.is_macos_or_newer` public API

malfet committed:
[MPS] Add MacOS 14 runtime check

Prerequisite for adding more complex type support and FFT operation

Check using `conjugateWithTensor:name:` selector defined as follows
```objc
/// Returns the complex conjugate of the input tensor elements.
///
/// - Parameters:
///   - tensor: The input tensor.
///   - name: An optional string which serves as an identifier for the operation..
/// - Returns: A valid `MPSGraphTensor` object containing the elementwise result of the applied operation.
-(MPSGraphTensor *) conjugateWithTensor:(MPSGraphTensor *) tensor
                                   name:(NSString * _Nullable) name
MPS_AVAILABLE_STARTING(macos(14.0), ios(17.0), tvos(17.0))
MPS_SWIFT_NAME( conjugate(tensor:name:) );
```

Rename `isOnMacOS13orNewer` hook to `isOnMacOSorNewer` as well as `torch._C.__mps_is_on_macos_13_or_newer` to `torch._C._mps_is_on_macos_or_newer`
Add `torch.backends.mps.is_macos_or_newer` public API

[ghstack-poisoned]

malfet committed:
Update on "[MPS] Add MacOS 14 runtime check"


Prerequisite for adding more complex type support and FFT operation

Check using `conjugateWithTensor:name:` selector defined as follows
```objc
/// Returns the complex conjugate of the input tensor elements.
///
/// - Parameters:
///   - tensor: The input tensor.
///   - name: An optional string which serves as an identifier for the operation..
/// - Returns: A valid `MPSGraphTensor` object containing the elementwise result of the applied operation.
-(MPSGraphTensor *) conjugateWithTensor:(MPSGraphTensor *) tensor
                                   name:(NSString * _Nullable) name
MPS_AVAILABLE_STARTING(macos(14.0), ios(17.0), tvos(17.0))
MPS_SWIFT_NAME( conjugate(tensor:name:) );
```

- Rename `isOnMacOS13orNewer(unsigned minor)` hook to `isOnMacOSorNewer(major, minor)`
- Replace `torch._C.__mps_is_on_macos_13_or_newer` with `torch._C._mps_is_on_macos_or_newer`
- Add `torch.backends.mps.is_macos_or_newer` public API

[ghstack-poisoned]

albanD commented on aten/src/ATen/mps/MPSHooks.mm:
```diff
@@ -19,7 +19,16 @@
+    } else if (minor > 0) {
+      TORCH_WARN("Can't check whether running on 14.", minor, "+ returning one for 14.0+");
+    }
+    return is_macos_13_or_newer(MacOSVersion::MACOS_VER_14_0_PLUS);
```
Should is_macos_13_or_newer be renamed as well?

malfet replied:
This is part of the stable API, and `is_macos_13_or_newer(MACOS_VER_14_0_PLUS)` sounds like a reasonable API call, but sure, I can rename it

albanD commented on torch/backends/mps/__init__.py:
```diff
@@ -22,10 +22,16 @@ def is_available() -> bool:
 
+@_lru_cache
+def is_macos_or_newer(major: int = 13, minor: int = 0) -> bool:
+    r"""Return a bool indicating whether MPS is running on gven MacOS or newer."""
```
```suggestion
    r"""Return a bool indicating whether MPS is running on given MacOS or newer."""
```

albanD commented on torch/backends/mps/__init__.py:
```diff
@@ -22,10 +22,16 @@ def is_available() -> bool:
 
 
+@_lru_cache
+def is_macos_or_newer(major: int = 13, minor: int = 0) -> bool:
```
Do you really want default values here?

malfet committed:
Update on "[MPS] Add MacOS 14 runtime check"


Prerequisite for adding more complex type support and FFT operation

Check using `conjugateWithTensor:name:` selector defined as follows
```objc
/// Returns the complex conjugate of the input tensor elements.
///
/// - Parameters:
///   - tensor: The input tensor.
///   - name: An optional string which serves as an identifier for the operation..
/// - Returns: A valid `MPSGraphTensor` object containing the elementwise result of the applied operation.
-(MPSGraphTensor *) conjugateWithTensor:(MPSGraphTensor *) tensor
                                   name:(NSString * _Nullable) name
MPS_AVAILABLE_STARTING(macos(14.0), ios(17.0), tvos(17.0))
MPS_SWIFT_NAME( conjugate(tensor:name:) );
```

- Rename `isOnMacOS13orNewer(unsigned minor)` hook to `isOnMacOSorNewer(major, minor)`
- Replace `torch._C.__mps_is_on_macos_13_or_newer` with `torch._C._mps_is_on_macos_or_newer`
- Add `torch.backends.mps.is_macos_or_newer` public API

[ghstack-poisoned]

albanD reviewed (approved):
SGTM!

malfet commented:
@pytorchbot merge -f "Lint + MacOS binary builds are green"
